### DISCUSSION TITLE
How to work with clangd?

### DISCUSSION THREAD

**Discussion by CherryPetale** [mkslanc/ace-linters#92] [Q&A]
I would like to use clangd to get intellisense of C, but I am not sure how to implement it, so I would like to ask a few questions.

* **console error**
  I got somewhat close to completion, but when I finally called the following `languageProvider.format();`, an error occurred. This may not be all that clear, but do you have any idea what it is?
https://github.com/mkslanc/ace-linters/blob/953b3ff248e847edb9e9d48873afe26bc1788385/packages/demo/websockets-lsp/client.ts#L33-L41

   ```
   Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'reverse')
       at ace-language-client.js:20924:36
       at MessageController.callbackFunction (ace-language-client.js:19422:17)
       at MessageController.emit (ace-language-client.js:2504:5)
       at MockWorker.<anonymous> (ace-language-client.js:19434:18)
       at MockWorker.emit (ace-language-client.js:2504:5)
       at MockWorker.postMessage (ace-language-client.js:20975:27)
       at MockWorker.<anonymous> (ace-language-client.js:10019:17)
       at MockWorker.emit (ace-language-client.js:2504:5)
       at MockWorker.postMessage (ace-language-client.js:20975:27)
       at MessageController.postMessage (ace-language-client.js:19426:22)
   ```

* **support for `stdio`**

  I don't think ace-linters currently support `stdio`, so I had to convert the clangd input/output to web socket.
  I searched and found a similar question on [stackoverflow](https://stackoverflow.com/questions/73924659/how-to-start-clangd-via-web-socket) and followed the answers on this site to set up a server using `jsonrpc-ws-proxy.`
  The connection seems to be working, but I am not sure if this approach is really correct.
  Do you plan to support `stdio`?

* **official support for clangd**
   I would like to see official support for the clangd. Is there a possibility?

PS: These texts are translated. Please note that this is poor English.

**Comment by mkslanc** [comment]
Hi, @CherryPetale !
- console error 
 It seems worked for me though using my [websockets bridge](https://github.com/mkslanc/websockets-ls/blob/main/server.js#L54C27-L54C27). I added example for clang there.
- Yes, currently it is not supported, but we have in plans to add `stdio` support
- When stdio/ipc would be added there shouldn't be any problems with clangd, except (maybe) some server configuration issues

**Comment by CherryPetale** [comment]
@mkslanc  Thanks for the reply.

I decided to verify it once with monaco editor to identify the problem since you say it works fine. Then it worked fine in monaco editor.
Therefore, I found that the problem was on the ace editor side, so I investigated and found that the `modes` was incorrectly specified.
The ace editor also initiated communication with the server, but error `failed to parse URI at (root).textDocument.uri` was displayed.
So I referred to the sample code again and found that there is a process to convert uri.
https://github.com/mkslanc/websockets-ls/blob/c63e29f566004830288538ecd7b8c3f7bb1f8114/server.js#L113-L118

This was not necessary in monaco editor, but is this process required?

**Reply by mkslanc** [reply]
I would say, it depends of the use case, right now I assume that it's server's (or bridge's) responsibility to resolve uri's correctly, cause in browser env we don't have normal access to file system

**Reply by CherryPetale** [reply]
I see. thank you for your reply.
